console-subscriber/src/visitors.rs

```diff
@@ -22,15 +23,21 @@ const LOCATION_COLUMN: &str = "loc.col";
 ///     "runtime.resource",
 ///     concrete_type = "Sleep",
 ///     kind = "timer",
+///     is_internal = true,
+///     inherits_child_attrs = true,
 /// );
 ///
 /// Fields:
 /// concrete_type - indicates the concrete rust type for this resource
 /// kind - indicates the type of resource (i.e. timer, sync, io )
+/// is_internal - whether this is a resource type that is not exposed publicly (i.e. BatchSemaphore)
+/// inherits_child_attrs - whether this resource should inherit the state attributes of its children
```

Comment:
hmm, the fact that we have a sort of magic field that controls how the console displays data is...not exactly my favorite thing? it seems like in an ideal world, the tracing events would _just_ expose data, and the console would be able to determine how it should be displayed, and this kind of feels like a violation of that idea. but, i also get that maybe this is necessary to get the data to display nicely.

is there any other potential solution that would let us get the UI we want without having the runtime annotate its events with this field? because...that would certainly be nice.

Reply:
@hawkw If you are referring to `inherits_child_attrs` I totally agree. The problem is that in some scenarios we want to propagate these attrs to the parent resource/async op. For example in the situation of Semaphore/BatchSempahore. Most of the instrumentation points that update state are happening in the BatchSempahore but we want the same data on the public Semaphore as it will be a bit hard to instrument that one properly. One possible solution that will most likely yield a similar result to what we have at the moment is to make these state attrs always be inherited from child resources that are flagged as internal. In that scenario however, we will have the `Mutex` resource inherit the state attrs of the internal Semaphore that it uses. Of course we can think of some way to differentiate them in the UI to make it clear that these are attrs of an internal resource but again... this can become confusing. Do you have any better ideas?

Reply:
> The problem is that in some scenarios we want to propagate these attrs to the parent resource/async op. For example in the situation of Semaphore/BatchSempahore. Most of the instrumentation points that update state are happening in the BatchSempahore but we want the same data on the public Semaphore as it will be a bit hard to instrument that one properly. 

yeah, i get the motivation for this, and it is entirely possible that it _is_ the best solution! 

one thing i might consider trying is a global toggle in the resources view that sets whether or not private resources are displayed. when private resources are turned on, we would show those resources in the resource list _and_ display any child attributes from private resources on the parent resources. when the toggle is turned off, we would hide both the private resources from the resource list, _and_ hide their attributes from parent resources. it could be worth playing with that and seeing if it's an acceptable solution to this problem?

if a toggle in the UI doesn't adequately solve the problem, i'd be okay with moving forward with this approach; it just seems worth trying?

Reply:
another thing to consider is, maybe we just always have parents inherit child attributes if they have no attributes of their own, regardless of whether or not the toggle is set?